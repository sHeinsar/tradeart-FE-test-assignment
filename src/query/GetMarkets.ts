import { gql } from '@apollo/client'

export const GET_MARKETS = gql`
    query($ticker: String!) {
        markets(filter:{ marketSymbol: {_eq:$ticker} }) {
            baseSymbol
            ticker {
                lastPrice
            }
        }
    }
`
