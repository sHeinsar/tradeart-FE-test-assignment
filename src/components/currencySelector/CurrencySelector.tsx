import React, { useEffect, useState } from 'react'
import {
  Container,
  StyledInput,
  StyledButton,
  StyledTermsAndConditions,
  StyledInputText,
  InputContainer,
  LoadingIcon
} from './CurrencySelector.styled'
import { useLazyQuery } from '@apollo/client'
import { CurrencyData } from '../../types/cryptoCurrency'
import { GET_MARKETS } from '../../query/GetMarkets'

interface CurrencySelectorProps {
  handleAddCurrency: (currency: CurrencyData) => void
}

const CurrencySelector: React.FC<CurrencySelectorProps> = ({ handleAddCurrency }) => {
  const [getData, { data, loading }] = useLazyQuery(GET_MARKETS, { fetchPolicy: 'network-only' })
  const [error, setError] = useState(false)
  const [input, setInput] = useState('')

  useEffect(() => {
    if ((data?.markets && data.markets.length < 1) && input !== '') {
      setError(true)
      setInput('')
    }
    if ((data?.markets && data.markets.length > 0) && input !== '') {
      setError(false)
      handleAddCurrency({
        ticker: data.markets[0].baseSymbol,
        price: data.markets[0].ticker.lastPrice
      })
    }
  }, [data])

  return (
        <Container>
            <InputContainer>
                <StyledInput data-testid="input" value={input} hasError={error} onChange={(event) => { setInput(event.target.value) }}/>
                {loading
                  ? <LoadingIcon />
                  : <StyledInputText hasError={error}>
                        {`Cryptocurrency code ${error ? 'invalid' : ''}`}
                    </StyledInputText>
                }
            </InputContainer>
            <StyledButton
                disabledStyle={input === ''}
                disabled={input === ''}
                onClick={async () => {
                  if (input === '') {
                    setError(true)
                    return
                  }
                  await getData({ variables: { ticker: `Binance:${input}/EUR` } })
                  setInput('')
                }}>
                Add
            </StyledButton>
            <StyledTermsAndConditions>Use of this service is subject to terms and conditions</StyledTermsAndConditions>
        </Container>
  )
}

export default CurrencySelector
