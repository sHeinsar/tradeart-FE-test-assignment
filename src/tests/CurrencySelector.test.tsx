import { MockedProvider } from '@apollo/client/testing'
import { GET_MARKETS } from '../query/GetMarkets'
import { fireEvent, render, waitFor, screen } from '@testing-library/react'
import CurrencySelector from '../components/currencySelector/CurrencySelector'
import React from 'react'

test('component renders correctly', () => {
  const handleAddCurrency = jest.fn()
  const mocks = [
    {
      request: {
        query: GET_MARKETS,
        variables: { ticker: 'Binance:BTC/EUR' }
      },
      result: {
        data: {
          markets: [
            {
              baseSymbol: 'BTC',
              ticker: {
                lastPrice: '12345'
              }
            }
          ]
        }
      }
    }
  ]
  const { container } = render(
        <MockedProvider mocks={mocks}>
            <CurrencySelector handleAddCurrency={handleAddCurrency}/>
        </MockedProvider>
  )
  expect(container.firstChild).toMatchSnapshot()
  expect(container.querySelector('input')).toBeTruthy()
  expect(container.querySelector('button')).toBeTruthy()
  expect(handleAddCurrency).toHaveBeenCalledTimes(0)
})

test('handleAddCurrency function is called when an invalid input is entered and the "Add" button is clicked', async () => {
  const handleAddCurrency = jest.fn()
  const mocks = [
    {
      request: {
        query: GET_MARKETS,
        variables: { ticker: 'Binance:BTC/EUR' }
      },
      result: {
        data: {
          markets: [
            {
              baseSymbol: 'BTC',
              ticker: {
                lastPrice: '12345'
              }
            }
          ]
        }
      }
    }
  ]
  const { getByText } = render(
      <MockedProvider mocks={mocks}>
        <CurrencySelector handleAddCurrency={handleAddCurrency}/>
      </MockedProvider>
  )

  const button = getByText(/Add/i)

  fireEvent.click(button)
  await waitFor(() => {
    expect(handleAddCurrency).toHaveBeenCalledTimes(0)
  })
})

test('handleAddCurrency function is called when an valid input is entered and the "Add" button is clicked', async () => {
  const handleAddCurrency = jest.fn()
  const mocks = [
    {
      request: {
        query: GET_MARKETS,
        variables: { ticker: 'Binance:BTC/EUR' }
      },
      result: {
        data: {
          markets: [
            {
              baseSymbol: 'BTC',
              ticker: {
                lastPrice: '12345'
              }
            }
          ]
        }
      }
    }
  ]
  const { getByText } = render(
      <MockedProvider mocks={mocks}>
        <CurrencySelector handleAddCurrency={handleAddCurrency}/>
      </MockedProvider>
  )

  const button = getByText(/Add/i)
  const input = screen.getByTestId('input')

  fireEvent.change(input, { target: { value: 'BTC' } })
  fireEvent.click(button)
  await waitFor(() => {
    expect(handleAddCurrency).toHaveBeenCalledTimes(1)
  })
})
