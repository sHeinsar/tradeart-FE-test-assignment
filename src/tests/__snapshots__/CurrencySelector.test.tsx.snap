// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders correctly 1`] = `
<div
  class="sc-bcXHqe bPUpPz"
>
  <div
    class="sc-gswNZR bAnNyH"
  >
    <input
      class="sc-dkrFOg hBIVuz"
      data-testid="input"
      value=""
    />
    <div
      class="sc-hLBbgP bZskFg"
    >
      Cryptocurrency code 
    </div>
  </div>
  <button
    class="sc-eDvSVe jBGzVC"
    disabled=""
  >
    Add
  </button>
  <span
    class="sc-jSUZER iWOEvo"
  >
    Use of this service is subject to terms and conditions
  </span>
</div>
`;
